### PR TITLE
[Bugfix] Make metadata send non-blocking in GroupCoordinator.isend_tensor_dict

### PR DESCRIPTION
**Pull request by RookieCoder-Camera**
## Purpose
---
  
  Two blocking/ownership defects in GroupCoordinator.isend_tensor_dict:

 Bug 1 — blocking metadata send. isend_tensor_dict is nominally non-blocking, but it delivered the metadata via a synchronous send_object (gloo send). gloo sends above the eager threshold are rendezvous: they block until the peer posts the matching recv. Pipeline schedules that post a send whose matching recv only happens in a later iteration (e.g. virtual-pipeline fold-point sends in vllm-ascend) deadlock both ranks inside the metadata send — the exact opposite of what isend_ promises.

 Bug 2 — no retention for fire-and-forget callers. The pickled metadata buffers (size_tensor / object_tensor) are local temporaries; if the caller drops the returned handles, nothing keeps them alive while gloo copies them on a background thread → use-after-free mid-send. Likewise, the device tensor sources get record_stream only when tensor.is_cuda, so on non-CUDA devices an in-flight isend buffer can be freed or overwritten by subsequent compute.

I expect a better performance when this bug is fixed for PP with chunked Prefill as shown below:
isend_tensor_dict_async_metadata

## Fix
---
  - Add isend_object(obj, dst) -> Handle: posts size + pickled object with isend (preserving gloo FIFO order) and returns a _RetainedHandle that owns the serialized CPU tensors until wait() drains both works.
  - isend_tensor_dict seeds its handle list with isend_object(...) instead of the blocking send_object(...). send_tensor_dict still waits every handle, so its synchronous semantics are unchanged.
  - Self-retention for fire-and-forget callers: each isend_tensor_dict call appends (handles, sent_tensors) to a _pending_isends FIFO, reaped lazily on the next call. The reap gates only on the device-tensor handles (is_completed() is reliable there) and wait()s the gloo metadata handle as a backstop — never the reverse, because gloo Work.is_completed() can report completion before the background copy finishes. Metadata-only entries (no reliable completion signal) are dropped best-effort. The sync send_tensor_dict path pops its own entry after waiting, so no refs leak.

## Not a duplicate
  Checked open PRs touching this area: <fill in — run gh pr list --repo vllm-project/vllm --state open --search "isend_tensor_dict" and --search "send_object parallel_state">. This change is confined to GroupCoordinator.send/isend_tensor_dict + new isend_object/_RetainedHandle; no overlap with
  <...>.

## Test Plan
---
  New unit tests in tests/distributed/test_comm_ops.py: isend_object posts size-then-object and releases on wait; isend_tensor_dict returns the composite metadata handle; fire-and-forget self-retention + FIFO reap + metadata-only drop; sync send_tensor_dict leaves no retention entry.

  python -m pytest tests/distributed/test_comm_ops.py -k "isend or send_tensor_dict"
  6 passed

Additionally verified on 4 * 5090 GPU with PP*2 and TP*2 for Qwen3-32B. 
1、Accuracy Test(Pass):
| dataset | version | metric | mode | vllm-api-stream-chat |
|----- | ----- | ----- | ----- | -----|
| gsm8k | 50861a | accuracy | gen | 96.74 |
---

2、ThroughPut Test:
run 100 test cases with 2 prompts:
vllm bench serve   --host localhost   --port 9090   --model /data/weight/Qwen3-32B   --served-model-name qwen3_32b   --dataset-name random   --random-input-len 16384   --random-output-len 1   --num-prompts 2   --request-rate inf   --max-concurrency 2
ThroughPut with sync:
Total token throughput (tok/s):          6370.36
ThroughPut with async:
Total token throughput (tok/s):          6705.80

with 5.2% better performance

  No API/schema changes: isend_tensor_dict's return contract (caller may wait() the handles) is preserved; existing callers (gpu_worker._pp_send_work, send_tensor_dict) already wait and are unaffected.

## Disclosure
---
This fix was developed with AI assistance (Claude). I reviewed every changed line and ran the tests above myself.
Signed off by:
Xinyi Zhou, zhouxinyi6@huawei.com
Jingchun Gao，gaojingchun1@huawei.com

### CODE REVIEW

**Comment by njhill** [comment]
Thanks @RookieCoder-Camera.

Could you share the env vars / cmd line options used to run vllm for these benchmarks.

In particular we're only interested in the performance improvement with model runner v2 (`VLLM_USE_V2_MODEL_RUNNER=1`). This will be be made the default very soon (probably today).

PP is already known to be suboptimal with MRV1.

**Comment by njhill** [comment]
@RookieCoder-Camera I was able to repro similar improvements on 4x GB200:


| Config | PP2+TP2 base | PP2+TP2 PR | Delta | PP4 base | PP4 PR | Delta |
|---|---|---|---|---|---|---|
| A: 16k in / 1 out, 2 prompts, conc 2 (author's exact config) | 29.1k | 42.9k | **+47%** | 17.1k | 28.2k | **+65%** |
| B: 16k in / 1 out, 64 prompts, conc 8 (sustained prefill) | 57.9k | 57.6k | -0.5% | 59.4k | 61.2k | **+3.0%** |
| C: 4k in / 128 out, 64 prompts, conc 8 (prefill+decode mix) | 18.04k | 18.19k | +0.9% | 11.9k | 12.6k | **+5.4%** |

File: vllm/distributed/parallel_state.py [hunk 1]
```diff
@@ -882,6 +922,33 @@ def recv_object(self, src: int) -> Any:
...
+            "Invalid destination rank. Destination rank is the same "
+            "as the current rank."
+        )
+
+        object_tensor = torch.frombuffer(pickle.dumps(obj), dtype=torch.uint8)
+        size_tensor = torch.tensor(
+            [object_tensor.numel()], dtype=torch.long, device="cpu"
+        )
+
+        retained = [size_tensor, object_tensor]
```
Comment (njhill) [inline vllm/distributed/parallel_state.py]: nit: make this a tuple
```suggestion
        retained = (size_tensor, object_tensor)
```

File: vllm/distributed/parallel_state.py [hunk 2]
```diff
@@ -1070,6 +1193,19 @@ def isend_tensor_dict(
...
             handles.append(handle)
+            sent_tensors.append(tensor)
+
+        # Self-retain for fire-and-forget callers: keep the handles and
+        # the source tensors alive until the sends complete (reaped
+        # lazily by ``_reap_completed_isends``). Without this, dropping
+        # the returned handles would GC the ``_RetainedHandle``'s
+        # serialized metadata buffers mid-send, and non-CUDA devices
+        # have no ``record_stream`` equivalent protecting the tensor
+        # sources.
```
Comment (njhill) [inline vllm/distributed/parallel_state.py:1205]: can make this less wordy
```suggestion
        # lazily by ``_reap_completed_isends``).
```

File: vllm/config/vllm.py
Comment (njhill) [inline vllm/config/vllm.py]: Revert this

**Review by njhill — COMMENTED** [review]
Thanks @RookieCoder-Camera! Could you also merge in latest main and resolve conflicts

File: vllm/distributed/parallel_state.py (hunk 1 above)
Comment (RookieCoder-Camera) [reply vllm/distributed/parallel_state.py]: thanks, I've changed this line.

File: vllm/distributed/parallel_state.py (hunk 2 above)
Comment (RookieCoder-Camera) [reply vllm/distributed/parallel_state.py:1205]: thanks, I've made this less wordy

File: vllm/config/vllm.py
Comment (RookieCoder-Camera) [reply vllm/config/vllm.py]: OK, I have reverted the change in vllm.py. This change was originally made for MRV1.

**Review by njhill — APPROVED** [review]
